test/model/message_test.dart

```diff
@@ -76,6 +94,303 @@ void main() {
+      outboxMessage = store.outboxMessages.values.single;
+      check(outboxMessage).state.equals(OutboxMessageState.hidden);
+    }
+
+    late Future<void> failToSendMessageFuture;
+    Future<void> prepareSendMessageToFail({
+      Duration delay = Duration.zero,
+    }) async {
+      message = eg.streamMessage(stream: stream);
+      await prepare(stream: stream);
+      await prepareMessages([eg.streamMessage(stream: stream)]);
+      connection.prepare(apiException: eg.apiBadRequest(), delay: delay);
+      failToSendMessageFuture = store.sendMessage(
+        destination: streamDestination, content: 'content');
+      outboxMessage = store.outboxMessages.values.single;
+      check(outboxMessage).state.equals(OutboxMessageState.hidden);
+    }
+
+    test('smoke DM: hidden -> waiting -> waitPeriodExpired -> (delete)', () => awaitFakeAsync((async) async {
+      await prepareSendMessageToSucceed(destination: DmDestination(
+        userIds: [eg.selfUser.userId, eg.otherUser.userId]));
+      check(outboxMessage).state.equals(OutboxMessageState.hidden);
+
+      async.elapse(kLocalEchoDebounceDuration);
+      check(outboxMessage).state.equals(OutboxMessageState.waiting);
+
+      async.elapse(kSendMessageOfferRestoreWaitPeriod - kLocalEchoDebounceDuration);
+      check(outboxMessage).state.equals(OutboxMessageState.waitPeriodExpired);
+
+      await store.handleEvent(eg.messageEvent(
+        eg.dmMessage(from: eg.selfUser, to: [eg.otherUser]),
+        localMessageId: outboxMessage.localMessageId));
+      check(store.outboxMessages).isEmpty();
+    }));
+
+    test('smoke stream message: hidden -> waiting -> waitPeriodExpired -> (delete)', () => awaitFakeAsync((async) async {
+      await prepareSendMessageToSucceed();
+      check(outboxMessage).state.equals(OutboxMessageState.hidden);
+
+      async.elapse(kLocalEchoDebounceDuration);
+      check(outboxMessage).state.equals(OutboxMessageState.waiting);
+
+      async.elapse(kSendMessageOfferRestoreWaitPeriod - kLocalEchoDebounceDuration);
+      check(outboxMessage).state.equals(OutboxMessageState.waitPeriodExpired);
+
+      await store.handleEvent(eg.messageEvent(message,
+        localMessageId: outboxMessage.localMessageId));
+      check(store.outboxMessages).isEmpty();
+    }));
+
+    group('… -> failed', () {
+      test('hidden -> failed', () => awaitFakeAsync((async) async {
+        await prepareSendMessageToFail();
+        await check(failToSendMessageFuture).throws();
+        check(outboxMessage).state.equals(OutboxMessageState.failed);
+
+        // Wait till we reach at least [kSendMessageOfferRestoreWaitPeriod] after
+        // the send request was initiated.
+        async.elapse(kSendMessageOfferRestoreWaitPeriod);
+        async.flushTimers();
+        // The outbox message should stay in the failed state;
+        // it should not transition to waiting or waitPeriodExpired.
+        check(outboxMessage).state.equals(OutboxMessageState.failed);
+      }));
```

Comment:
I see your reply at https://github.com/zulip/zulip-flutter/pull/1472#issuecomment-2892391723 :

> I agree that checking the state after the outbox message is removed from the store does not seem necessary. Removed those checks, and kept this late variable is mostly a helper for us to check the state of the outbox message.

It's actually unhelpful to use the `late outboxMessage` variable here, isn't it (?), because if there's a bug where the outbox message is removed from the store when it shouldn't be, this test won't catch it.

If all the uses of `outboxMessage` were replaced with lookups in the store (via a helper function if needed), I think that would be better, because we'd be testing the interface of this subsystem more directly.

Reply:
> because if there's a bug where the outbox message is removed from the store when it shouldn't be, this test won't catch it

Oh right. That makes sense. Yeah, that will make it more end-to-end.